From ac9ab745d318a7e52368a943b03ec56eb6bde3ea Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:53:22 -0400
Subject: [tools] perf/symbols: Destroy unused symsrcs

Message-id: <1408953210-26343-315-git-send-email-jolsa@redhat.com>
Patchwork-id: 89024
O-Subject: [PATCH RHEL7.1 BZ1133083 314/322] perf symbols: Destroy unused symsrcs
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 98e9f03bbf2cb21a60f94b8b700eb5d38470819d
Author: Namhyung Kim <namhyung@kernel.org>
Date: Thu Feb 20 10:32:54 2014 +0900

description
===========
Stephane reported that perf report and annotate failed to process data
using lots of (> 500) shared libraries.  It was because of the limit on
number of open files (ulimit -n).

Currently when perf loads a DSO, it'll look for normal and dynamic
symbol tables.  And if it fails to find out both tables, it'll iterate
all of possible symtab types.  But many of them are useless since they
have no additional information and the problem is that it's not closing
those files even though they're not used.  Fix it.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/symbol.c b/tools/perf/util/symbol.c
index a9d758a..e89afc0 100644
--- a/tools/perf/util/symbol.c
+++ b/tools/perf/util/symbol.c
@@ -1336,6 +1336,8 @@ int dso__load(struct dso *dso, struct map *map, symbol_filter_t filter)
 
    if (syms_ss && runtime_ss)
     break;
+  } else {
+   symsrc__destroy(ss);
   }
 
  }
-- 
1.7.1